editor/src/clj/editor/collection_data.clj

```diff
@@ -1,33 +1,361 @@
-    (update string-encoded-embedded-component-data :data string->component-data)))
-
-(defn string-decode-game-object-data [ext->resource-type string-encoded-game-object-data]
-  (let [string-decode-embedded-component-data (partial string-decode-embedded-component-data ext->resource-type)
-        string-decode-embedded-components-data (partial mapv string-decode-embedded-component-data)]
-    (update string-encoded-game-object-data :embedded-components string-decode-embedded-components-data)))
-
-(defn string-decode-embedded-instance-data [ext->resource-type string-encoded-embedded-instance-data]
-  (let [game-object-resource-type (ext->resource-type "go")
-        game-object-read-fn (:read-fn game-object-resource-type)
-        string-decode-game-object-data (partial string-decode-game-object-data ext->resource-type)
-        string-decode-embedded-game-object-data (comp string-decode-game-object-data
-                                                      (fn [^String embedded-game-object-string]
-                                                        (with-open [reader (StringReader. embedded-game-object-string)]
-                                                          (game-object-read-fn reader))))]
-    (update string-encoded-embedded-instance-data :data string-decode-embedded-game-object-data)))
-
-(defn string-decode-collection-data [ext->resource-type string-encoded-collection-data]
-  (let [string-decode-embedded-instance-data (partial string-decode-embedded-instance-data ext->resource-type)
-        string-decode-embedded-instances-data (partial mapv string-decode-embedded-instance-data)]
-    (update string-encoded-collection-data :embedded-instances string-decode-embedded-instances-data)))
+(defn- file-not-found-error [referencing-node-id referencing-resource referenced-proj-path]
+  (let [referenced-resource (workspace/resolve-resource referencing-resource referenced-proj-path)]
+    (resource-io/file-not-found-error referencing-node-id nil :fatal referenced-resource)))
+
+(g/defnk produce-proj-path->build-target [referenced-component-build-targets referenced-collection-build-targets referenced-game-object-build-targets resource-property-build-targets]
+  (bt/make-proj-path->build-target
+    (concat referenced-collection-build-targets
+            referenced-game-object-build-targets
+            referenced-component-build-targets
+            resource-property-build-targets)))
+
+(defn- any-instance-desc->transform-matrix
+  ^Matrix4d [{:keys [position rotation scale3] :as any-instance-desc}]
+  ;; GameObject$InstanceDesc, GameObject$EmbeddedInstanceDesc, or GameObject$CollectionInstanceDesc in map format.
+  (let [corrected-scale (if (or (nil? scale3)
+                                (protobuf/default-read-scale-value? scale3))
+                          (or (:scale any-instance-desc) 1.0) ; Legacy file format - use uniform scale.
+                          scale3)]
+    (math/clj->mat4 position rotation corrected-scale)))
```

Comment:
`math/clj->mat4` supports both uniform scale as a number, or non-uniform scale as a Clojure vector of numbers.